### DISCUSSION TITLE
How can I pass broadcast metadata to a webhook in Ant Media Server for authentication and event notifications?

### DISCUSSION THREAD

**Discussion by Mohit-3196** [ant-media/Ant-Media-Server#5254] [FAQs]
I'm utilizing webhooks in Ant Media Server to handle stream authentication and receive notifications for events like `liveStreamStarted` and `liveStreamEnded`, etc.
In the [release logs](https://github.com/ant-media/Ant-Media-Server/releases) of version 2.6.0, it was mentioned that it's now possible to pass broadcast metadata to webhooks for authentication purposes.

Could anyone please guide me on the steps or instructions to accomplish this?

**Comment by lastpeony** [comment]
Enabling the webhook authentication feature allows the server to send an HTTP request to your provided webhook address whenever a stream is initiated for publishing. This request includes essential stream details such as the stream name, app name, and stream ID. Starting from version 2.6.0, this information also includes broadcast metadata. When you publish a stream, the metadata you have passed to the Ant Media server (which can be done through the .publish method in all SDKs or through REST API) is included.

Upon receiving this request, you have the opportunity to parse and process the information on your end and send a response. If the response code is 200, the server will authorize the stream and allow it to commence publishing. Conversely, if the response code differs from 200, the server will reject the stream from being published.

For more information about webhook authentication check this page on our documentation:
https://antmedia.io/docs/guides/advanced-usage/webhook-stream-authorization/

**Comment by rahul78275** [comment]
Hello @Mohit-3196, @lastpeony , @yashtandon113 

I tested the webhook feature on my local system and is working fine for events like (liveStreamStarted, liveStreamEnded). Thanks to the team for this great feature.

I have a question : 
Shall I get notification when I create a new RTMP endpoint using rest api, also when RTMP endpoint gets deleted because of original stream  stopped ?

Scenario : 
I am developing an application where a publisher have to publish a live stream to multiple social media platform, and I want to verify that all my endpoints are working fine. If any RTMP endpoint stops, my application should get notification so that i can show the publisher that your broadcast stopped for (lets say facebook, or youtube).

One approach is to call the rest api in every X interval of time and verify the existence. (This is not a preferable approach to me).

**Reply by Ramjivan** [reply]
+1 for this

**Comment by rahul78275** [comment]
Hello hello.. I am waiting 😃

**Comment by Mohit-3196** [comment]
Hi @rahul78275 ,
How are you doing?

I think stream created webhook notification is not available yet. But we may add this depending upon the demand. What do you think @lastpeony?
For the RTMP endpoint fails, there is a webhook call `endpointFailed` and you can check it from https://antmedia.io/docs/guides/advanced-usage/webhooks/

Please let me know if it doesn't help.

Thank you

**Reply by rahul78275** [reply]
Hello @Mohit-3196 abi 😄 

I am doing good, Thanks for asking.

I do not need stream created webhook notification (my mistake i wrote "Create" in my question). 

I got the solution while testing the apis. When i publish a webrtc stream from Javascript SDK, I got the webhook notification as (action=liveStreamStarted&id=streamId_BafFBzjQ6)

After then I created a new rtmp stream using api (/v2/broadcasts/{id}/rtmp-endpoint). Here i also got a webhook notification as (action=liveStreamStarted&id=testStreamRtmp)

Finally I stopped the primary stream (streamId_BafFBzjQ6). Here I received two webhook notification as expected
(action=liveStreamEnded&id=testStreamRtmp && action=liveStreamEnded&id=streamId_BafFBzjQ6)

So my observation is webhook notification is working in webrtc as well rtmp stream when it start/end.

I also noticed that by mistake in api payload for key "rtmpUrl", if i entered "http" in place of "rtmp", api returned success : true, 
It should return false with an error message.

![image](https://github.com/ant-media/Ant-Media-Server/assets/36676540/fc250461-ccec-485a-af09-2bd8afcf7f15)

Thank you
